include/aspect/particle/property/cpo_elastic_tensor.h

```diff
@@ -0,0 +1,204 @@
+/*
+ Copyright (C) 2022 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _aspect_particle_property_cpo_elastic_tensor_h
+#define _aspect_particle_property_cpo_elastic_tensor_h
+
+#include <aspect/particle/property/interface.h>
+#include <aspect/particle/property/crystal_preferred_orientation.h>
+#include <aspect/simulator_access.h>
+#include <array>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+
+      /**
+       * Computes the elastic tensor $C_{ijkl} based on the cpo in both olivine
+       * and enstatite. It uses a Voigt average.
+       *
+       * The layout of the data vector per partcle is the following (note that
+       * for this plugin the following dim's are always 3):
+       * 1 unrolled tensor -> 3x3x3x3 (dim*dim*dim*dim) doubles, starts at:
+       *                                   data_position
+       *
+       * @ingroup ParticleProperties
+       */
+      template <int dim>
+      class CpoElasticTensor : public Interface<dim>, public ::aspect::SimulatorAccess<dim>
+      {
+        public:
+          /**
+           * Constructor
+           */
+          CpoElasticTensor();
+
+          /**
+           * Initialization function. This function is called once at the
+           * beginning of the program after parse_parameters is run.
+           */
+          void
+          initialize () override;
+
+
+          /**
+           * This implements the Voigt averaging as described in the equation at the
+           * bottom of page 385 in Mainprice (1990, doi: 10.1016/0098-3004(90)90072-2):
+           * $C^V_{ijkl} = \sum^S_l F_s \sum^{N_s}_l C_{ijkl}/N_s$, where $F_s$ is the
+           * grain size, $N_s$ is the number of grains and $C_{ijkl}$ is the elastic
+           * tensor. This elastic tensor is computed by the equation above in
+           * Mainprice (1990): $C_{ijkl} = R_{ip} R_{jg} R_{kr} R_{is} C_{pgrs}$, where
+           * R_{ij} is the cpo orientation matrix.
+           */
+          virtual
```

Comment:
Does this function have to be virtual? Does it have to be public? If not, make it not virtual or not public.

Reply:
It doesn't have to be virtual, but it does have to be public. We could maybe move it to the new Utilities tensor space in a follow up pull request.